### ISSUE TITLE
Latest Composer is incompatible with latest Magento

### ISSUE THREAD

**Issue by nanderson11** [markshust/docker-magento#1084]
**Description**
When setting docker-magento up for the first time, it installs the latest version of Composer from what I see here https://github.com/markshust/docker-magento/blob/69f29dea57f12e8f7363e483b753a299d15cb030/images/php/8.2/Dockerfile#L95. Currently, this installs Composer 2.7.1. After setup, if I run `bin/composer` (without arguments), it outputs an error (below). This seems to be caused by the issue discussed here https://github.com/composer/composer/issues/11843, caused by an incompatibility with Composer 2.7+ and Magento.

**Expected Result**
I think it would be nice if instead of installing the latest version of Composer, it installed the latest version of Composer that's compatible with Magento as outlined here: https://experienceleague.adobe.com/docs/commerce-operations/installation-guide/system-requirements.html. I know that you can technically install different versions of Magento in docker-magento though, so maybe it just defaults to whatever version of Composer is compatible with the default version of Magento that's installed.

**Actual Result**
```
~/magento.test$ bin/composer
Fatal error: Declaration of Magento\ComposerRootUpdatePlugin\Plugin\Commands\OverrideRequireCommand::execute(Symfony\Component\Console\Input\InputInterface $input, Symfony\Component\Console\Output\OutputInterface $output) must be compatible with Composer\Command\RequireCommand::execute(Symfony\Component\Console\Input\InputInterface $input, Symfony\Component\Console\Output\OutputInterface $output): int in /var/www/html/vendor/magento/composer-root-update-plugin/Plugin/Commands/OverrideRequireCommand.php on line 173
```

**Comment by markshust** [comment]
Great catch -- it should definitely be pegged to the supported version of Composer for the most recent version, which at this time is 2.2.

**Comment by YevhenZvieriev** [comment]
I pegged the version of Composer for `2.2.0` for PHP 8.2

Should I do this for 8.1 and 8.3?

**Comment by YevhenZvieriev** [comment]
Hi, @markshust  

The [official system requirements for Magento 2.4.5](https://experienceleague.adobe.com/docs/commerce-operations/installation-guide/system-requirements.html) state that Composer 2.2 is supported.

However, Composer `2.2.0` throws the error 

```
Fatal error: Uncaught Error: Class "Magento\Framework\Component\ComponentRegistrar" not found in /var/www/html/setup/src/Magento/Setup/registration.php:9
```

during composer commands.

![Screenshot_2024-03-16_00-51-58](https://github.com/markshust/docker-magento/assets/43544955/9450161e-b32a-46c3-8619-404a3b4d3916)

This bug has been fixed in later versions of Composer 2.2. 

So, we should peg Composer for the `2.2.18` instead of the `2.2.0` version to resolve this issue.

**Comment by markshust** [comment]
Fixed in https://github.com/markshust/docker-magento/pull/1090. I'll tag new versions and recompile very shortly.
